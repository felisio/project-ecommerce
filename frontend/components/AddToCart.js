import React, { Component } from 'react';
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag';

import { CURRENT_USER_QUERY } from './User'

const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
    }
  }
`;

class AddToCart extends Component {
  render() {
    const { id } = this.props
    return (
      <Mutation 
        refetchQueries={[{query: CURRENT_USER_QUERY}]}
        mutation={ADD_TO_CART_MUTATION} 
        variables={{ id }}
      >
        {(addToCart, { loading }) => (
          <button disabled={loading} onClick={addToCart}>
            Add{loading && 'ing'} to Cart ðŸ›’
          </button>
        )}
      </Mutation>
    );
  }
}

export default AddToCart;
export { ADD_TO_CART_MUTATION }