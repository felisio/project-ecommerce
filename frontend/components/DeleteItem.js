import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag'

import { ALL_ITEMS_QUERY } from './Items'

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION ( $id: ID! ) {
    deleteItem( id: $id ) {
      id
    }
  }
`;

class DeleteItem extends Component {
  
  deleteItem = async(deleteItemMutation) => {
    if(confirm('Are you sure you want delete this item?')) {
      deleteItemMutation().catch(err => {
        alert(err.message)
      });
    }
  }

  update = (cache, payload) => {
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY })
    data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id)
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data })
  }

  render() {
    return (
      <Mutation 
        refetchQueries={['ALL_ITEMS_QUERY', 'PAGINATION_QUERY']}
        mutation={DELETE_ITEM_MUTATION} 
        variables={{ id: this.props.id }}
        update={this.update}
      >
        {(deleteItem, {loading, error}) => (
          <button onClick={() => this.deleteItem(deleteItem)}>{this.props.children}</button>
        )}
      </Mutation>
    );
  }
}

export default DeleteItem;