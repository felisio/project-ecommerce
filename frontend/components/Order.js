import React, { Component } from 'react';
import PropTypes from 'prop-types'
import { Query } from 'react-apollo'
import { format } from 'date-fns'
import Head from 'next/head'
import gql from 'graphql-tag'
import formatMoney from '../lib/formatMoney'
import Error from './ErrorMessage'
import OrderStyles from './styles/OrderStyles'

const SINGLE_ORDER_QUERY = gql`
  query SINGLE_ORDER_QUERY($id: ID!) {
    order(id: $id) {
      id
      charge
      total
      createAt
      user {
        id
      }
      items {
        id
        title
        description
        image
        price
        quantity
      }
    }
  }
`;

class Order extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
  }

  render() {
    const { id } = this.props;
    return (
      <Query query={SINGLE_ORDER_QUERY} variables={{id}}>
        {({data, loading, error}) => {
          if(error) return <Error error={error} />
          if(loading) return <p>Loading...</p>
          const { order } = data
          return (
            <OrderStyles>
              <Head>
                <title>Sick Fit order - {order.id}</title>
              </Head>
              <p>
                <span>Order Id</span>
                <span>{order.id}</span>
              </p>
              <p>
                <span>Charge</span>
                <span>{order.charge}</span>
              </p>
              <p>
                <span>Date</span>
                <span>{format(order.createAt, 'MMMM d, YYYY h:mm a')}</span>
              </p>
              <p>
                <span>Total</span>
                <span>{formatMoney(order.total)}</span>
              </p>
              <p>
                <span>Items Count</span>
                <span>{order.items.length}</span>
              </p>
              <div className="items">
                {order.items.map(item => (
                  <div className="order-item" key={item.id}>
                    <img src={item.image} alt={item.title}/>
                    <div className="item-details">
                      <h2>{item.title}</h2>
                      <p>Qtd: {item.quantity}</p>
                      <p>Price: {formatMoney(item.price * item.quantity)}</p>
                      <p>{item.description}</p>
                    </div>
                  </div>
                ))}
              </div>
            </OrderStyles>
          )
        }}
      </Query>
    );
  }
}

export default Order;