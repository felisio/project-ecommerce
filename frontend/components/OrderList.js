import React, { Component } from 'react';
import PropTypes from 'prop-types'
import { Query } from 'react-apollo'
import { formatDistance } from 'date-fns'
import Link from 'next/link'
import Head from 'next/head'
import gql from 'graphql-tag'
import formatMoney from '../lib/formatMoney'
import Error from './ErrorMessage'
import styled from 'styled-components'
import OrderItemStyles from './styles/OrderItemStyles';

const LIST_ORDER_QUERY = gql`
  query LIST_ORDER_QUERY {
    orders(orderBy: createAt_DESC) {
      id
      charge
      total
      createAt
      items {
        id
        title
        description
        image
        price
        quantity
      }
    }
  }
`;

const OrderUl = styled.ul`
  display: grid;
  grid-gap: 4rem;
  grid-template-columns: repeat(auto-fit, minmax(40%, 1fr));
`;

class OrderList extends Component {
  render() {
    return (
      <Query query={LIST_ORDER_QUERY}>
        {({ data, error, loading }) => {
          if(error) return <Error error={error} />
          if(loading) return <p>Loading...</p>
          const { orders } = data

          return(
            <div>
              <h2>You have {orders.length} Orders</h2>
              <OrderUl>
                {orders.map(order => (
                  <OrderItemStyles key={order.id}>
                      <Link href={{ pathname: '/order', query: {id: order.id} }}>
                        <a>
                          <div className="order-meta">
                            <p>{order.items.reduce((a, b)=> a + b.quantity, 0)} Items</p>
                            <p>{order.items.length} Products</p>
                            <p>Create: {formatDistance(order.createAt, new Date())}</p>
                            <p>Total: {formatMoney(order.total)}</p>
                          </div>
                          {/* <div className="images">
                            {order.items.map(item => (
                              <img src={item.image} alt={item.title} key={item.id}/>
                            ))}
                          </div> */}
                        </a>
                      </Link>
                  </OrderItemStyles>
                ))}
              </OrderUl>
            </div>
          );
        }}
      </Query>
    );
  }
}

export default OrderList;