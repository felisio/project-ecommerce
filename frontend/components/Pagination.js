import React from 'react'
import { Query } from 'react-apollo'
import gql from 'graphql-tag'
import Head from 'next/head'
import Link from 'next/link'

import PaginationStyles from './styles/PaginationStyles'
import { perPage } from '../config'

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`;

const Pagination = props => (
  <Query query={PAGINATION_QUERY}>
      {({data, error, loading}) => {
        if (loading) return <p>Loading...</p>
        if (error) return <p>Error: {error.message}</p>
        const count = data.itemsConnection.aggregate.count
        const pages = Math.ceil(count / perPage)
        const { page } = props
        return (
          <PaginationStyles data-test="pagination">
            <Head>
              <title>Sick Fits | page {page} of {pages}</title>
            </Head>
            <Link
              prefetch 
              href={{
                pathname: 'items',
                query: { page: page - 1 },
              }}
            >
              <a className="prev" aria-disabled={page <= 1}> ðŸ”™ Prev </a>
            </Link>
            <p>
              {page} of 
              <span className="total-pages">
                {pages}
              </span>
            </p>
            <p>{count} items total!</p>
            <Link href={{
              pathname: 'items',
              query: { page: page + 1 },
            }}>
              <a className="next" aria-disabled={page >= pages}> Next ðŸ”œ </a>
            </Link>
          </PaginationStyles>
        )
      }}
    </Query>
)

export default Pagination
export { PAGINATION_QUERY };