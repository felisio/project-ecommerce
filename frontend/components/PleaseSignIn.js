import { Query } from 'react-apollo'
import gql from 'graphql-tag'

import Signin from './Signin'
import { CURRENT_USER_QUERY } from './User'

const PleaseSignIn = props => (
  <Query query={CURRENT_USER_QUERY}>
    {({data, loading}) => {
      if (loading) return <p>Loading...</p>
      if(!data.me) {
        return (
          <div>
            <h2>Please Sign in!</h2>
            <Signin />
          </div>
        )
      }
      return props.children
    }}
  </Query>
)

export default PleaseSignIn