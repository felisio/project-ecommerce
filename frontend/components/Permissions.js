import { Query } from 'react-apollo'
import gql from 'graphql-tag'
import Table from './styles/Table'

import Error from './ErrorMessage'
import Signin from './Signin'
import { CURRENT_USER_QUERY } from './User'
import PermissionsItem from './PermissionsItem'

export const possiblePermissions = [
  'ADMIN',
  'USER',
  'ITEMCREATE',
  'ITEMUPDATE',
  'ITEMDELETE',
  'PERMISSIONUPDATE',
];

const ALL_USERS_QUERY = gql`
query ALL_USERS_QUERY {
  users {
    id
    name
    email
    permissions
  }
}
`;

const Permissions = props => (
  <Query query={ALL_USERS_QUERY}>
    {({data, loading, error}) => {
      if (error) return <Error error={error} />
      if (loading) return <p>Loading...</p>
      if(data.users) {
        return (
          <div>
            <h2>Manage Permissions</h2>
            <Table>
              <thead>
                <tr>
                  <th>Name</th>
                  <th>Email</th>
                  {possiblePermissions.map(permission => <th key={permission}>{permission}</th>)}
                  <th>ğŸ‘‡</th>
                </tr>
              </thead>
              <tbody>
                {data.users.map(user => <PermissionsItem key={user.id} user={user} />)}
              </tbody>
            </Table>
          </div>
        )
      }
      return props.children
    }}
  </Query>
)

export default Permissions