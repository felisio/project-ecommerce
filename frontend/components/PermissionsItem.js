import React from 'react';
import PropTypes from 'prop-types'
import { Mutation } from 'react-apollo'

import SickButton from './styles/SickButton'
import { possiblePermissions } from './Permissions'
import {gql} from 'apollo-boost';

const UPDATE_PERMISSION_MUTATIONS = gql`
  mutation UPDATE_PERMISSION_MUTATIONS( $permissions: [Permission], $userId: ID! ) {
    updatePermissions( permissions: $permissions, userId: $userId ) {
      id
      name
      email
      permissions
    }
  }
`

class PermissionsItem extends React.Component {
  static propTypes = {
    user: PropTypes.shape({
      id: PropTypes.string,
      email: PropTypes.string,
      name: PropTypes.string,
      permissions: PropTypes.array,
    }).isRequired
  }

  state = {
    permissions: this.props.user.permissions
  }

  handlePermissionChange = (e, updatePermissionsMutation) => {
    const checkbox = e.target
    let updatePermissions = [...this.state.permissions]
    if(checkbox.checked) {
      updatePermissions.push(checkbox.value)
    }else {
      updatePermissions = updatePermissions.filter(item => item !== checkbox.value)
    }
    this.setState({ permissions: updatePermissions }, updatePermissionsMutation)
  }

  render() {
    const { user } = this.props
    return (
      <Mutation 
        mutation={UPDATE_PERMISSION_MUTATIONS}
        variables={{
          permissions: this.state.permissions,
          userId: user.id,
        }}
      >
        {(updatePermissions, {loading, error}) => (
          <tr key={user.id}>
            <td>{user.name}</td>
            <td>{user.email}</td>
            {possiblePermissions.map(permission => (
              <td key={permission}>
                <label htmlFor={`${user.id}-permission-${permission}`}>
                  <input 
                    id={`${user.id}-permission-${permission}`}
                    type="checkbox"
                    checked={this.state.permissions.includes(permission)} 
                    value={permission}
                    onChange={(e) => this.handlePermissionChange(e, updatePermissions)}
                  />
                </label>
              </td>
            ))}
            <td>
              <SickButton type="button" disabled={loading} onClick={updatePermissions}>
                Updat{loading ? 'ing' : 'e'}
              </SickButton>
            </td>
          </tr>
        )}
      </Mutation>
    )
  }
}

export default PermissionsItem