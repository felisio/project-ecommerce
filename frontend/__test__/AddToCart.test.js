import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils'
import { ApolloConsumer } from 'react-apollo'

import { fakeCartItem, fakeUser } from '../lib/testUtils'
import AddToCart, { ADD_TO_CART_MUTATION } from '../components/AddToCart'
import { CURRENT_USER_QUERY } from '../components/User';

const cartItem = fakeCartItem();

const mock = [
  // current user query
  {
    request: { query: CURRENT_USER_QUERY },
    result: { 
      data: { 
        me: {
          ...fakeUser(),
          cart: [], 
        }
      },
    },
  },
  {
    request: { query: ADD_TO_CART_MUTATION, variables: { id: '123' } },
    result: { 
      data: {
        addToCart: {
          ...cartItem,
          quantity: 1,
        }
      } 
    },
  },
  {
    request: { query: CURRENT_USER_QUERY },
    result: { 
      data: { 
        me: {
          ...fakeUser(),
          cart: [cartItem], 
        }
      },
    },
  },
];

describe('<AddToCart />', () => {
  it('renders and macth snapshot', async() => {
    const wrapper = mount(
      <MockedProvider mocks={mock}>
        <AddToCart id="123" />
      </MockedProvider>
    );

    await wait();
    wrapper.update();
    expect(toJson(wrapper.find('button'))).toMatchSnapshot() 
  })

  it('add to cart is clicked', async() => {
    let apolloClient;
    const wrapper = mount(
      <MockedProvider mocks={mock}>
        <ApolloConsumer>
          {(client) => {
            apolloClient = client
            return  <AddToCart id="123" />
          }}
        </ApolloConsumer>
      </MockedProvider>
    );

    await wait();
    wrapper.update();
    const { data: { me } } = await apolloClient.query({ query: CURRENT_USER_QUERY })
    expect(me.cart).toHaveLength(0);
    // add item
    wrapper.find('button').simulate('click');
    await wait();
    const { data: { me: me2 } } = await apolloClient.query({ query: CURRENT_USER_QUERY })
    expect(me2.cart).toHaveLength(1);
    expect(me2.cart[0].id).toBe('omg123');
  })
})