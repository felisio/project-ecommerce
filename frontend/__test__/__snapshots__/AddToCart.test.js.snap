// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddToCart /> renders and macth snapshot 1`] = `
<button
  disabled={false}
  onClick={[Function]}
>
  Add
   to Cart ðŸ›’
</button>
`;
